### PR TITLE
metadata: process metadata from incoming image frames

### PR DESCRIPTION
**Pull request by deadprogram**
This PR adds the feature of parsing metadata such as the status of the I/O lines and LVAL count from an attached encoder.

### CODE REVIEW

File: examples/metadata/main.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+	fmt.Printf("\r\n %s      %s      %s      %s      %s      %s",
+		showValue(md.IIN1()),
+		showValue(md.IIN2()),
+		showValue(md.IIN3()),
+		showValue(md.IIN4()),
+		showValue(md.DIN1()),
+		showValue(md.DIN2()))
+	fmt.Printf("\r\n------------------------------------------------->")
+	fmt.Println()
+}
```
Comment (julian59189) [inline examples/metadata/main.go:196]: should this be part of metadata.go?

File: examples/metadata/main.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+	fmt.Println("x:", x, "y:", y, "pitch:", pitch)
+
+	// turn on metadata insertion.
+	if err := ch.SetParamInt(mc.MetadataInsertionParam, mc.MetadataInsertionEnable); err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	// Retrieve channel metadatacontent information.
+	content, err = ch.GetParamInt(mc.MetadataContentParam)
```
Comment (julian59189) [inline examples/metadata/main.go:126]: What is content getting set to here? What does it depend on?

**Review by julian59189 — APPROVED** [review]
Looks just. Just had a few questions about things that where not clear to me

File: examples/metadata/main.go (hunk 2 above)
Comment (deadprogram) [reply examples/metadata/main.go:126]: It is based on the hardware configuration itself, not settable in software.

File: examples/metadata/main.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+		fmt.Println(err)
+		return
+	}
+	defer ch.Delete()
+
+	SetupCamera()
+
+	ch.RegisterCallback(cbhandler)
+
+	// MC_SIG_SURFACE_PROCESSING: acquisition done and locked for processing
```
[the remark this answers is not in the document]
Comment (deadprogram) [reply examples/metadata/main.go:47]: Yes, that means successful acquisition of a frame has been completed.

File: examples/metadata/main.go (hunk 1 above)
Comment (deadprogram) [reply examples/metadata/main.go:196]: Maybe? it is useful.

File: examples/metadata/main.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+		return
+	}
+
+	fmt.Println("metadatacontent:", content)
+
+	// The number of images to acquire.
+	if err := ch.SetParamInt(mc.SeqLengthFrParam, mc.IndeterminateLength); err != nil {
+		fmt.Println(err)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (deadprogram) [reply examples/metadata/main.go:138]: This says "just keep capturing" vs. "capture one frame/capture exactly N frames". They call that "indeterminate length" of number of frames.

File: examples/metadata/main.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (deadprogram) [reply examples/metadata/main.go:47]: That event is "a single frame has now been completely acquired, so the frame data is available".

File: examples/metadata/main.go (hunk 1 above)
Comment (deadprogram) [reply examples/metadata/main.go]: Implemented as `Stringer` interface so `fmt.Println()` "just works".

**Comment by deadprogram** [comment]
OK I think all comments have now been addressed.

**Pull request merged**